### ISSUE TITLE
Bug: MemoryLeak in 2.1.1 

### ISSUE THREAD

**Issue by olzzon** [nrkno/sofie-casparcg-server#6]
I have found a memory leak in 2.1RC2 loop files on 4 outputs and calling INFO 3 times every second. 
Still investigating.
Has this been fixed since that version.?
Otherwise I will update and see if I can reproduce it on newest version.

**Comment by Julusian** [comment]
No, I'm not aware of any memory leaks and haven't looked into any recently.

How fast is it leaking?

If you can narrow it down to being clips on loop or calling INFO frequently that would also be helpful to know but not essential :)

**Comment by olzzon** [comment]
Thanks.
I just wanted to check before seeing up a test. I can reproduce it, but it´s over a couple of days (I found it as I was long time testing my cog-state-scanner) and that´s why I wanted to be sure that it was not found before testing :)
I´ll give some feedback about it after the weekend. And report if it´s INFO, filetype, fps, loop, decklink or something else related :)

**Comment by olzzon** [comment]
Just an intermediate status. (before upgrading to newer version etc.)
Please chime in if someone can reproduce. I´ll try it on another machine and with newest version 2.1.1 the next couple of days. To locate it further.

I can reproduce it with the go1080p25 clip. on a machine with no ACMP call except for the 4 commands:
```
play 1-10 go1080p25 loop
play 2-10 go1080p25 loop
play 3-10 go1080p25 loop
play 4-10 go1080p25 loop
```

No other ACMP commands (e.g. INFO) has been called, and no other software is running on the machine.

It´s CCG 2.1RC2 running on Windows 10 with a decklink Quad2 card.
Machine: Intel i7, 16gig ram, Nvidia 1050 graphics card.
And the usage from CCG has gone from around 900mb to 1,7gb in 8 hours. (and will keep growing)

´´´
<img width="723" alt="image" src="https://user-images.githubusercontent.com/29782499/47928276-0bf07900-dec6-11e8-9055-154471b34076.png">
´´´

**Comment by olzzon** [comment]
I can replicate it on a Xeon with no decklink and just 3 <screen> output like this:
```
<channel>
      <video-mode>1080i5000</video-mode>
      <channel-layout>stereo</channel-layout>
      <consumers>
        <screen>
          <device>1</device>
          <windowed>true</windowed>
        </screen>
        <system-audio></system-audio>
      </consumers>
    </channel>
    <channel>
```

Both machines are set to 1080i5000, so it might be related to 25/50i, as I´m using the go1080p25 file. 
Will try with other files :)

**Comment by olzzon** [comment]
Before upgrading I tried the CG1080i50 file.
It does the same with:
```
play 1-10 CG1080i50 loop
play 2-10 CG1080i50 loop
play 3-10 CG1080i50 loop
play 4-10 CG1080i50 loop
```
So it´s not related to 25/50i, now it´s time to try other CCG versions.
I´ll be back :)

**Comment by Julusian** [comment]
Great, that is a very simple STR to follow and gives a good idea of where to start looking.
Do you know if doing a `CLEAR 1-10` or `CLEAR 1` will free the leaked memory?

At this point I think I have as much information as I need to take a look into this.

Thanks for your help and all the info :)

**Comment by olzzon** [comment]
Hi Julian, 
Thanks for the hint. 
I´m in the middle of testing the machine that has the problem with <screen> in stead of <decklink> just to be sure that it´s related to decklink output.
After that, probably this evening, I´ll try to test the CLEAR command.

**Comment by olzzon** [comment]
No memory leak when outputting to screen instead of Decklink. Now on for testing the CLEAR 1-10 command :)

**Comment by jesperstarkar** [comment]
Perhaps I should join and test 4 channels just running html templates on decklink outputs then.

**Comment by Julusian** [comment]
I am pretty certain that key-device won't do anything as its not a number so is invalid. custom-allocator is not a valid option either.

I don't expect there are any significant differences in the rest of the config, so don't mind which we choose to go with

**Comment by jesperstarkar** [comment]
> key-device won't do anything as its not a number so is invalid

It's too early for me -> getting some coffee.

**Comment by jesperstarkar** [comment]
![screenshot 2018-11-06 at 11 29 52](https://user-images.githubusercontent.com/6236874/48058738-7f8cd180-e1b7-11e8-9bbb-087ee9b0a123.png)
<img width="858" alt="screenshot 2018-11-06 at 10 29 55" src="https://user-images.githubusercontent.com/6236874/48058739-7f8cd180-e1b7-11e8-970f-9f9bf7226f24.png">
![screenshot 2018-11-06 at 09 48 17](https://user-images.githubusercontent.com/6236874/48058740-80256800-e1b7-11e8-82d4-0e1b04101c6f.png)

**Comment by jesperstarkar** [comment]
- NRK Rc2
- Decklink drivers 10.10
- Quad2

```
<?xml version="1.0"?>
<!-- Generated with SuperSuite by SuperFly.tv -->
<configuration>
  <paths>
    <media-path>c:/CasparCG/media</media-path>
    <log-path>c:/CasparCG/log/</log-path>
    <data-path>c:/CasparCG/data/</data-path>
    <template-path>c:/CasparCG/template/</template-path>
    <thumbnail-path>thumbnail/</thumbnail-path>
    <font-path>font/</font-path>
  </paths>
  <channels>
    <channel>
      <video-mode>1080i5000</video-mode>
      <consumers>
     	<decklink>
      		<device>1</device>
      		<embedded-audio>true</embedded-audio>
      	</decklink> 
      </consumers>
    </channel>
    <channel>
      <video-mode>1080i5000</video-mode>
      <consumers>
      	<decklink>
      		<device>2</device>
      		<embedded-audio>true</embedded-audio>
      	</decklink>
      </consumers>
    </channel>
    <channel>
      <video-mode>1080i5000</video-mode>
      <consumers>
      	<decklink>
      		<device>3</device>
      		<embedded-audio>true</embedded-audio>
      	</decklink>
      </consumers>
    </channel>
    <channel>
      <video-mode>1080i5000</video-mode>
      <consumers>
      <decklink>
      		<device>4</device>
      		<embedded-audio>true</embedded-audio>
      	</decklink>
      </consumers>
    </channel>
  </channels>
  <controllers>
    <tcp>
      <port>5250</port>
      <protocol>AMCP</protocol>
    </tcp>
  </controllers>
   <osc>
    <default-port>6250</default-port>
    <disable-send-to-amcp-clients>false</disable-send-to-amcp-clients>
  </osc>
</configuration>
```

**Comment by jesperstarkar** [comment]
After clear (rock stable):

![screenshot 2018-11-06 at 11 34 10](https://user-images.githubusercontent.com/6236874/48058882-ef9b5780-e1b7-11e8-8e99-89bc9eefbe47.png)

**Comment by jesperstarkar** [comment]
**Findings:**
1) HTML producer doesn't seem to leak
2) FFmpeg producer seems to only leak when producing frames. Playing "amb" without loop stops growth after the last frame freezes. This indicates that the problem is in the producer not Decklink consumer.
3) Gonna try a long file without looping to see if it behaves the same as with looping.

**Comment by olzzon** [comment]
I can reproduce it on 2.1.1 also

**Comment by jesperstarkar** [comment]
I'm gonna try 2.2b7

**Comment by jesperstarkar** [comment]
No leaks in 2.2.x

**Comment by olzzon** [comment]
Changed the title, so it reflect our findings.

**Comment by olzzon** [comment]
Has anyone looked further into this one?

**Comment by olzzon** [comment]
Just to follow up on this one.

**Comment by olzzon** [comment]
Just to follow up on this one

**Comment by olzzon** [comment]
Just to follow up on this one.
I´ve had CasparCG 2.1.0 NRK RC2 running for more than a month, with wipes, html overlay all controlled from ClipTool. Without a single crash.
The config had only screenproducer running and no DeckLink in config.
Has anyone tried running it with NDI only or another VideoCard and experienced crashes?

**Comment by jesperstarkar** [comment]
Would be tempting to have you add decklink then.

**Comment by olzzon** [comment]
Yep, thats my plan. Just wanted to see if someone else had progress in this.

**Comment by Julusian** [comment]
I'm currently doing decklink in and out and it appears to be leaking a bit. I have an idea of where that is. I'll try and fix that after lunch then try looping ffmpeg again

**Comment by Julusian** [comment]
I believe I have fixed the cause of this leak. It was the decklink timecode output values not being tracked properly. 
After fixing that I got an increase of <1mb over an hour instead of ~50mb when doing 4xgo1080p25 on loop.

**Comment by olzzon** [comment]
Soooo coool :)
When there a build with the fix, I´ll install it for long term testing.

**Comment by jesperstarkar** [comment]
@JonStaleCarlsen 

Just waiting for a build from @Julusian

**Comment by Julusian** [comment]
Sorry about the delay. Test build: https://drive.google.com/open?id=1n-TCaqtI82LSgF0d4SP9B61KSlNTS9eu

**Comment by olzzon** [comment]
Great.
I have it running on 2 machines, testing hopefully for a couple of days :)

**Comment by olzzon** [comment]
So far things looks good :)

**Comment by olzzon** [comment]
I think you nailed it.
Everything still plays on both machines :)

**Comment by olzzon** [comment]
Running for a week now, so far so good.it looks like there´s still a minor leak somewhere, but I´ll keep it running to see how long it will play.
<img width="531" alt="image" src="https://user-images.githubusercontent.com/29782499/57679293-39fc0600-762b-11e9-84b4-d86e9a0d70df.png">

**Comment by Julusian** [comment]
That is great to hear :)
Hopefully that minor leak is really minor because finding it will not be an easy task...

**Comment by Julusian** [comment]
This should be long fixed, and will not be a problem in the 2.4 builds
